### ISSUE TITLE
AccessViolationException when logger and ares is active

### ISSUE THREAD

**Issue by Falco20019** [grpc/grpc#18074]
### What version of gRPC and what language are you using?
gRPC 1.18.0 for C#
 
### What operating system (Linux, Windows, …) and version?
Windows 18.03 in a company network with domain
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
Visual Studio 2017 / MSBuild
 
### What did you do?
```csharp
using Grpc.Core;
using Grpc.Core.Logging;
using Helloworld;
using System;
using System.IO;
using System.Threading.Tasks;

namespace Crashtest
{
    class Program
    {
        public static void Main()
        {
            string hostname = Environment.MachineName; // any non-localhost
            int port = 8010;

            Environment.SetEnvironmentVariable("GRPC_TRACE", "cares_resolver");
            Environment.SetEnvironmentVariable("GRPC_VERBOSITY", "DEBUG");
            Environment.SetEnvironmentVariable("GRPC_DNS_RESOLVER", "ares");

            GrpcEnvironment.SetLogger(new TextWriterLogger(TextWriter.Synchronized(File.CreateText("test.txt"))));

            var server = new Server
            {
                Services = { Greeter.BindService(new HelloWorldImpl()) },
                Ports = { new ServerPort(hostname, port, ServerCredentials.Insecure) }
            };
            server.Start();

            var client = new Greeter.GreeterClient(new Channel(hostname, port, ChannelCredentials.Insecure));
            var result = client.SayHello(new HelloRequest {Name = "Tester"}); // This will throw in GrpcThreadPool.RunHandlerLoop
        }

        internal class HelloWorldImpl : Greeter.GreeterBase
        {
            public override Task<HelloReply> SayHello(HelloRequest request, ServerCallContext context)
            {
                return Task.FromResult(new HelloReply {Message = "Hello " + request});
            }
        }
    }
}
```
 
### What did you expect to see?
The example should run fine and log messages to test.txt.
 
### What did you see instead?
An AccessViolationException will be thrown on the client call.
 
### Anything else we should know about your project / environment?
It works fine when not setting `GRPC_DNS_RESOLVER=ares`.

The `cares_resolver` tracer flag is missing in https://github.com/grpc/grpc/blob/master/doc/environment_variables.md but found it in the code base. The error also appears when using `all` as tracer.

**Comment by apolcyn** [comment]
Thanks for filing. Are you able to paste the logs here, including the exception?

**Comment by Falco20019** [comment]
It‘s sadly not flushing the full log and since it’s in a pooled thread, I was not able to catch it and close the application gracefully.

**Comment by apolcyn** [comment]
I haven't repro'd the access violation.

However, I can repro a name resolution - in any case one thing that this bug makes apparent is that c-ares can't resolve `Environment.MachineName`.

**Comment by apolcyn** [comment]
Ok, I can repro an access violation, specifically when I'm:
a) running under `GRPC_TRACE=all'
b) Configuring the grpc logger to log to a file, e.g.:

```
GrpcEnvironment.SetLogger(new TextWriterLogger(TextWriter.Synchronized(File.CreateText("test.txt"))));
```

**Comment by Falco20019** [comment]
Logging to another endpoint worked in one try. Haven’t tested it repeatedly. Even with this constellation, it’s not occurring everytime. But pretty regularly. It seems to be a combination of TextWriterLogger and the ares resolver. Even when using GRPC_TRACE=all, this only happens with GRPC_DNS_RESOLVER=ares.

**Comment by apolcyn** [comment]
Yeah, that "DNS resolution failure" that I thought I was encountering was actually just a subtle bug, described in https://github.com/grpc/grpc/pull/18084 - WIP fixing this.

**Comment by apolcyn** [comment]
FTR, regarding the "access violation": this too appears to be caused by same the fact that we make incorrectly error out of a DNS query because we a stale WSA socket error code comparison. When we do this, we initiate a WSA operation on a socket but [don't call grpc_socket_notify_on_read](https://github.com/grpc/grpc/blob/master/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver_windows.cc#L131). So when we then [destroy](https://github.com/grpc/grpc/blob/master/src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_ev_driver_windows.cc#L93) the socket, we'll [not realize ](https://github.com/grpc/grpc/blob/9963a1b87da3aae996d9a1d24bf95dc57349007e/src/core/lib/iomgr/socket_windows.cc#L101) that it has an ongoing I/O operation and so prematurely free it.

**Comment by apolcyn** [comment]
The fix for this was backported to 1.19.x in https://github.com/grpc/grpc/pull/18092, so it should be released in the 1.19.0 release soonish
